arangod/Aql/DocumentProducingHelper.h

```diff
@@ -154,191 +160,256 @@ struct DocumentProducingFunctionContext {
+  UniqueDocumentProducingFunctionContext(
+      InputAqlItemRow const& inputRow, OutputAqlItemRow* outputRow,
+      RegisterId const outputRegister, bool produceResult,
+      std::vector<std::string> const& projections, transaction::Methods* trxPtr,
+      std::vector<size_t> const& coveringIndexAttributePositions,
+      bool allowCoveringIndexOptimization, bool useRawDocumentPointers)
+      : DocumentProducingFunctionContext(inputRow, outputRow, outputRegister,
+                                         produceResult, projections, trxPtr,
+                                         coveringIndexAttributePositions,
+                                         allowCoveringIndexOptimization,
+                                         useRawDocumentPointers),
+        _alreadyReturned(),
+        _isLastIndex(false) {}
+
+  void reset() override {
+    _alreadyReturned.clear();
+    _isLastIndex = false;
+  }
+
+  bool checkUniqueness(LocalDocumentId const& token) override {
+    if (!_isLastIndex) {
+      // insert & check for duplicates in one go
+      if (!_alreadyReturned.insert(token.id()).second) {
+        // Document already in list. Skip this
+        return false;
+      }
+    } else {
+      // only check for duplicates
+      if (_alreadyReturned.find(token.id()) != _alreadyReturned.end()) {
+        // Document found, skip
+        return false;
+      }
+    }
+    return true;
+  }
+
+  void setIsLastIndex(bool val) override { _isLastIndex = val; }
+
+ private:
+  /// @brief set of already returned documents. Used to make the result distinct
+  std::unordered_set<TRI_voc_rid_t> _alreadyReturned;
+
+  /// @brief Flag if the current index pointer is the last of the list.
+  ///        Used in uniqueness checks.
+  bool _isLastIndex;
+};
+
 enum class DocumentProducingCallbackVariant {
-  NoResult,
   WithProjectionsCoveredByIndex,
   WithProjectionsNotCoveredByIndex,
   DocumentWithRawPointer,
   DocumentCopy
 };
 
-inline std::function<void(LocalDocumentId const&)> getNullCallback(DocumentProducingFunctionContext& context) {
-  return [&context](LocalDocumentId const&) {
-    InputAqlItemRow const& input = context.getInputRow();
-    OutputAqlItemRow& output = context.getOutputRow();
-    RegisterId registerId = context.getOutputRegister();
-    // TODO: optimize this within the register planning mechanism?
-    TRI_ASSERT(!output.isFull());
-    output.cloneValueInto(registerId, input, AqlValue(AqlValueHintNull()));
-    TRI_ASSERT(output.produced());
-    output.advanceRow();
-    context.incrScanned();
-  };
-}
-
-template<DocumentProducingCallbackVariant variant>
-inline DocumentProducingFunction getCallback(DocumentProducingFunctionContext& context);
+template <bool checkUniqueness>
+inline DocumentProducingFunction getCallback(DocumentProducingCallbackVariant const& variant,
+                                             DocumentProducingFunctionContext& context) {
```

Comment:
Alternatively, you could use tag types instead of an enum class and define separate overloads for the different callback variants (=tag dispatching):
```cpp
namespace DocumentProducingCallbackVariant  {
  struct WithProjectionsCoveredByIndex{};
  ...
}

template <bool checkUniqueness>
inline DocumentProducingFunction getCallback(
  DocumentProducingCallbackVariant::WithProjectionsCoveredByIndex,
  DocumentProducingFunctionContext& context) {
  //...
}
// and the same for the other variants...
```
This way you avoid the big switch, resulting in smaller functions for each variant. And in addition you get a compiler error if you use a variant without defining the corresponding overload.

Reply:
Nice suggestion, will try today.